Applications/Spire/Source/KeyBindings/KeyBindings.cpp

```diff
@@ -23,17 +79,21 @@ bool KeyBindings::CancelActionBinding::operator !=(
 KeyBindings KeyBindings::get_default_key_bindings() {
   auto bindings = KeyBindings();
   auto region = Region::Global();
-  auto action1 = KeyBindings::OrderAction{"", OrderType::LIMIT, Side::BID, {},
-    100, {}};
+  auto action1 = KeyBindings::OrderAction{"", {Tag(40, OrderType(
+    OrderType::LIMIT)), Tag(54, Side(Side::BID)), Tag(59,
+    boost::optional<TimeInForce>()), Tag(38, 100)}};
   bindings.set(Qt::Key_F5, region, action1);
-  auto action2 = KeyBindings::OrderAction{"", OrderType::LIMIT, Side::ASK, {},
-    100, {}};
+  auto action2 = KeyBindings::OrderAction{"", {Tag(40, OrderType(
```

Comment:
Unfortunately, the type of something like `OrderType::LIMIT` is technically not `OrderType`, it is `OrderTypeEnumMembers::Type` due to the definition of `BEAM_ENUM`. That makes `tag.get_value<OrderType>` fail should such a value be passed directly into the variant. It seems quite unsafe because one is running a risk of misinterpreting the value of a tag, but I do not know a better way to handle it and implicit conversions in general.

Reply:
Add overloaded constructor that takes an ```OrderType::Type``` and a ```Side::Type```.

Reply:
I think I can avoid API bloat using a slightly different approach. Instead of adding more constructors, I add such types into the variant type list, so that it is capable of storing both `OrderType` and `OrderTypeEnumMembers::Type`. To allow seamless casts like this:

```
auto tag = Tag(1, Side::BID);
auto val1 = tag.get_value<OrderType>(); // OK
auto val2 = tag.get_value<OrderTypeEnumMembers::Type>(); // OK
```

I then enable implicit (and optionally explicit as well) conversions in` Tag::get_value<T>` by converting `boost::optional<StoredType>` to `boost::optional<QueryType>`, which is easily done.

The side effect is then that queries like below will also succeed:

```
auto tag = Tag(1, 10);
auto value tag.get_value<double>();
```

which might be either desired or undesired depending on our preference. I think it is more user-friendly if any convertible (at least implicitly) type can be user as a tag's value during initialization, and the set of such types can change without the need to modify `KeyBindings`.

Reply:
Disregard the last part about modifying `KeyBindings`, one still has to add new convertible types to the variant. However, failing to do so will result in a compilation error, making it quite safe.

Reply:
I do like you're thinking on this issue and your solution is very friendly/permissive. Unfortunately I get worried about permissive and flexible code in C++ because without the right abstraction built directly into the language, permissive code in C++ almost always ends up breaking in a subtle and disasterous way at the worst possible time.

It will not be desirable to ```get_value<OrderType::Type>```, only ```get_value<OrderType>```. It also should not be permissible to store an ```int``` and get back a ```double```. Every tag is required to work with a specific type and may only work with that specific type.

You can add the constructors for ```Side::Type``` and ```OrderType::Type``` or you can require the user to have to wrap those values explicitly in which case ```Tag(1, Side::BID)``` is a compiler error, and it's not really the end of the world.

Reply:
Makes sense. The case with `Side` and `OrderType` is not that bad due to compiler errors.
What I think is more error-prone is something like `Quantity`, because both double and `Quantity` are accepted by the variant, but querying querying `Quantity` will fail if I store a `double`. In fact, I did this mistake already in `get_default_key_bindings` and test cases (which I forgot to commit, will include along with the fixes). I could implement tag dispatch or something to require explicit specification of a type, but this adds even more boilerplate and does not solve the problem completely. Anyway, with cast failures, it will at least manifest itself in runtime, so I think I will keep it as is and avoid adding extra constructors so that a user is prepared to explicitly cast the types on their own.